File: bin/array_processing/sigcor/get_tau.f
      subroutine get_tau(y1,y2,npts,sr,is,js,laglen,tau,ccc,cccp,
     &ireturn)      
c
c     This subroutine cross-correlates seismograms.  
c     It is assumed that waveforms are real*4 in y1 and y2.
c
c     y1,y2:       data series to be cross-correlated
c     npts:        length of the time series
c     sr:          sampling rates of the series
c     is,js:       station numbers for this cross-correlation
c     laglen:      maximum amount (seconds) of offset for cross-correlation
c                      laglen << npts/sr (laglen ~ 0.2 npts/sr is typical)
c     tau:         signal time diff. (epoch seconds) of first - second
c                  (This closely reflects the origin time difference.)
c     ccc:         cross-correlation coefficient (0 <= ccc <= 1)
c     cccp:        relative amplitude of first event over second event
c     ireturn:     error flag (0 means no error)
c                  1 -- real sample rate too far from integer value
c                  2 -- incompatible sample rates 
c                    (isr1 != isr2) && (isr1 != 2*isr2) && (isr2 != 2*isr1)
c                  3 -- number of requested lags is larger than dimensioned,
c                       so need to recompile with larger maxlag -- be sure 
c                       to change in all parameter statements
c
      parameter   (maxlag=1001)
      real        y1(*),y2(*)
      real        z(-maxlag:maxlag)
      real*8      sr,laglen,tau

c     tol is the tolerance to allow the real sampling rate to deviate from an
c     integer value; for instance for tol = 0.01, 99.99 and 100.01 are
c     acceptable and are taken to be nominally 100.  This value is set here
c     to 0.01, and will produce only up to 1/10 of a millisecond error in 1 sec.
      data        tol/0.01/

      ireturn = 0
c     write(7,'(a,2i4)') "i,j= ",is,js
c     write(7,*) y1(1),y2(1),y1(npts),y2(npts),npts

      lags = nint(laglen*sr)
c     write(7,*) ' sr,npts,lags = ',sr,npts,lags
      if (lags.gt.maxlag) then
        ireturn = 3
        return
      endif
      del = 1.0/sr

c     Get cross-correlation in time domain and normalize by standard deviations.
c     Compute the delay.  A positive tmax means the 2nd signal is delayed
c     wrt 1st by that amount.
c     The positive maximum (rmax) is returned, assuming the signals all have
c     the same polarity.
      call stats(y1,npts,2,mean1,stdev1,dummy,dummy)
      call stats(y2,npts,2,mean2,stdev2,dummy,dummy)
      call correl(y1,y2,z,npts,lags,rmax,kmax)
      ccc=rmax/(stdev1*stdev2)
      cccp=rmax/(stdev2*stdev2)
c     Occasional cases may arise where this leads to CCF > 1, so reset it 
c     to +1 if greater.
      if (ccc.gt. 1.0) ccc =  1.0
c     write(6,*) kmax,del
      tmax=kmax*del
      do k=-lags,lags
        z(k)=z(k)/(stdev1*stdev2)
      enddo
c     write(7,*) 'ccc,kmax = ',ccc,kmax
c     write(7,'(5e12.3)') (z(k),k=-lags,lags)
c     Pick out 3 points around maximum and compute best-fit parabola.
      ym=z(kmax-1)
      y0=z(kmax)
      yp=z(kmax+1)
      a = ( ym/2 + yp/2 - y0)/del**2
      b = (-ym/2 + yp/2)/del
      c = y0
c     Peak is at relative time where derivative = 0.
      tpeak = -b/(2*a)
c     Get total time shift.
      tau=tmax+tpeak
c     write(7,*) "tmax,tpeak,tau = ",tmax,tpeak,tau
c     If the 2nd is later than the 1st, then tau is positive.
      return
      end


      subroutine interp(x,n)
   
      real x(*)

      x(n*2-1) = x(n)
      do i = n-1,1,-1
        x(i*2-1) = x(i)
        x(i*2) = (x(i) + x(i+1))/2 
      enddo
      n = n*2 - 1

      return
      end

 
      subroutine correl(x,y,z,n,lags,rmax,kmax)
c
c     This returns the crosscorrelation function, normalized by the number
c     of points used but not by the stdev's of the two time series.
c
c     rmax is the maximum of the correlation function (positive). 
c     kmax is the point of this maximum
c     Note: The zeroth lag will be the midpoint of the returned series z.
c
      parameter   (maxlag=1001)
      real x(*),y(*),z(-maxlag:maxlag)
      rmax=0.0
      do j=-lags,lags
        z(j)=0.0
        do i=1,n
          if(i+j.ge.1 .and. i+j.le.n) z(j)=z(j)+x(i)*y(i+j)
        enddo
c       Normalize the CCF by the actual number of points used.  This accounts
c       for using less of the time series as the shift increases.  
        z(j)=z(j)/(n - abs(j))
        if (z(j).gt.rmax) then
          rmax=z(j)
          kmax=j
        endif
      enddo
      return
      end

      subroutine stats(x,n,k,amean,stdev,skew,kurtosis)
c
c     Computes statistics of a data vector.  How many stats is determined by
c     the value of k.
c
c     k = 1     mean
c     k = 2     also standard deviation
c     k = 3     also skew
c     k = 4     also kurtosis
c
      real*4 x(n)

      amean=0.0
      stdev=0.0
      skew=0.0
      kurtosis=0.0

      if (k.eq.0) return

      sum=0.0

      do i=1,n
        sum=sum+x(i)
      enddo

      amean=sum/n

      if (k.eq.1) return

      sumsq=0.0

      do i=1,n
        sumsq=sumsq+(x(i)-amean)**2
      enddo

      stdev=sqrt(sumsq/n)

      if (k.eq.2) return

      sumcu=0.0

      do i=1,n
        sumcu=sumcu+(x(i)-amean)**3
      enddo

      skew=(sumcu/n)/stdev**3

      if (k.eq.3) return

      sumqu=0.0

      do i=1,n
        sumqu=sumqu+(x(i)-amean)**4
      enddo

      kurtosis=(sumqu/n)/stdev**4

      return
      end

File: bin/array_processing/src/get_tau.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
 
int correl(float x[],float y[],float *z,int npts,int lags,float *rmax,int *kmax);
int nint(double x);
int stats(float x[], int npts, int k, float *mean, float *stdev, float *skew, float *kurtosis);

int get_tau(float y1[],float y2[],double tstart1,double tstart2,int npts,
double laglen,double sr,double *tau,float *ccc,float *cccp)
/*
      This function cross-correlates time series.  
      It is assumed that the time series are float arrays in y1 and y2.
      The additional array z is created to hold the cross-correlation series.
 
      y1,y2:       data series to be cross-correlated
      tstart1,2    start times (epoch seconds) of data series
      npts:        length of the two series
      sr:          sampling rate of the two series (not a whole number)
      laglen:      maximum amount (seconds) of offset for cross-correlation
                       laglen << npts/sr (laglen ~ 0.2 npts/sr is typical)
      tau:         signal time diff. (epoch seconds) of second - first
                   (This closely reflects the origin time difference.)
      ccc:         cross-correlation coefficient (-1 <= ccc <= 1)
      cccp:        relative amplitude of second event over first event

      return value:0 -- no error
                   1 -- real sample rate too far from integer value
                   2 -- number of requested lags is larger than length of series
                   3 -- divisor = 0.0 in computing refined tau
*/ 
{
  int         iret,len,isr,lags,kmax,i;
  float       *z;
  float       dummy,amean1,amean2,stdev1,stdev2,del,tmax,rmax;
  float       ym,y0,yp,a,b,c,tpeak,tshift;
  float       tol=0.01;
/*tol is the tolerance to allow the real sampling rate to deviate from an
  integer value; for instance for tol = 0.01, 99.99 and 100.01 are
  acceptable and are taken to be nominally 100.  This value is set here
  to 0.01, and will produce only up to 1/10 of a millisecond error in 1 sec.*/

/*Check if sample rate is nearly integer.  Some tolerance must be allowed 
  because sample rates given by CSS3.0 data may not be exactly whole 
  numbers.*/
  isr = nint(sr);
  if (fabs(isr - sr) > tol) return 1;

/*Get mean and stdev estimate and remove mean.*/
  stats(y1,npts,2,&amean1,&stdev1,&dummy,&dummy);
  for (i=0;i<npts;i++) y1[i] = y1[i] - amean1;
  stats(y2,npts,2,&amean2,&stdev2,&dummy,&dummy);
  for (i=0;i<npts;i++) y2[i] = y2[i] - amean2;

  lags = nint(laglen*sr);
  if (lags > npts) return 2;
  len = 2*lags + 1;
  z = malloc(len*sizeof(float));
  del = 1.0/sr;

/*Get cross-correlation in time domain and normalize by standard deviations.
  Normalize by stn. dev. squared of the first event to get an estimate
  of the amplitude of the second event relative to that of the first.  The
  log10 of this estimate would estimate the magnitude difference.*/

  iret = correl(y1,y2,z,npts,lags,&rmax,&kmax);

  *ccc  = rmax/(stdev1*stdev2);
  *cccp = rmax/(stdev1*stdev1);

/*Occasional cases may arise where this leads to |CCF| > 1, so reset it 
  to +/-1 if greater.  Such cases arise when the interpolated peak of the 
  cross-correlation function goes above +1 or below -1.*/
  if (*ccc >  1.0) *ccc =  1.0;
  if (*ccc < -1.0) *ccc = -1.0;

/*Compute the delay.  A positive tmax means the 2nd signal is delayed
  wrt 1st by that amount.*/
  tmax=kmax*del;
/*Pick out 3 points around maximum and compute best-fit parabola.*/
  ym = z[kmax+lags-1];
  y0 = z[kmax+lags];
  yp = z[kmax+lags+1];
  a = ( ym/2 + yp/2 - y0)/pow((double) del,(double)2);
  b = (-ym/2 + yp/2)/del;
  c = y0;
/*Peak is at relative time where derivative = 0.*/
  if (a == 0.0) 
  {
    free(z);
    return 3;
  }
  tpeak = -b/(2*a);
/*Get total time shift.*/
  tshift = tmax + tpeak;
/*Compute absolute time shift between signals.  If the 2nd is later than the 
  1st, the shift is positive.*/
  *tau = tstart2 - tstart1 + tshift;

  free(z);
  return 0;
}


int correl(float x[],float y[],float *z,int npts,int lags,float *rmax,int *kmax)
/*
      This returns the crosscorrelation function, normalized by the number
      of points used but not by the stdev's of the two time series.

      x,y  = two time series
      z    = array to hold the cross-correlation function
      npts = length of the two time series
      lags = number of lags  (n = 2*lags + 1)
      rmax = maximum of the correlation function (positive or negative). 
             To handle cases where polarity reversal may occur, the maximum is 
             allowed to be positive or negative.
      kmax = point of this maximum (relative to center point)
             e.g., if the zero-lag coef. is maximum, kmax = 0

      Note: The zeroth lag will be the midpoint of the returned series z.
*/ 
{
  int n,j,i;

  *rmax = 0.0;
  n = 2*lags + 1;
  for (j=0;j<n;j++)
  {
    z[j] = 0.0;
    for (i=0;i<npts;i++)
      if(i+j-lags >= 0 && i+j-lags < npts) z[j] = z[j] + x[i]*y[i+j-lags];
/*  Normalize the CCF by the full length.  This is like doing a circular
    cross-correlation with zero padding.*/
    z[j] = z[j]/npts;
    if (fabs(z[j]) > fabs(*rmax))
    {
      *rmax = z[j];
      *kmax = j-lags;
    }
  }
  return 0;
}
